File: 1268 - Search Suggestions System/solution.py
from bisect import bisect_left
class Solution:
    def suggestedProducts(self, A, s):
        A.sort()
        res, prefix, idx = [], '', 0
        for c in s:
            prefix += c
            idx = bisect_left(A, prefix, lo=idx)
            res.append([i for i in A[idx:idx+3] if i.startswith(prefix)])
        return res

File: 1268 - Search Suggestions System/solution.cpp
class Solution {
public:
    vector<vector<string>> suggestedProducts(vector<string>& A, string s) {
        int n = s.size(), idx = 0;
        string prefix = "";
        vector<vector<string>> res(n);
        sort(A.begin(), A.end());
        for(int i = 0; i < n; i++){
            prefix += s[i];
            idx = lower_bound(A.begin()+idx, A.end(), prefix)-A.begin();
            for(int j = idx; j < min(idx+3, (int)A.size()); j++){
                if(A[j].compare(0, i+1, prefix) == 0)
                    res[i].push_back(A[j]);
            }
        }
        return res;
    }
};